### PR TITLE
GPU support - GPU memory limitation and CUDA profiler

### PR DESCRIPTION
**Pull request by Raalsky**

### CODE REVIEW

**Review by grzegorz-gutowski — APPROVED** [review]

**Review by grzegorz-gutowski — APPROVED** [review]

**Review by grzegorz-gutowski — APPROVED** [review]

File: kolejka/judge/systems/local.py [hunk 1]
```diff
@@ -223,6 +276,10 @@ def execute_safe_command(self, command, stdin_path, stdout_path, stdout_append,
         resources = self.get_resources(limits)
         #resources[resource.RLIMIT_NPROC] = (1,1) #This is a very bad idea, read notes in man execv on EAGAIN
 
+        memory_reservation_thread = None
+        if limits.gpu_memory:
+            memory_reservation_thread = threading.Thread(target=memory_reservation, args=(limits.gpu_memory,))
+            memory_reservation_thread.start()
```
Comment (grzegorz-gutowski) [inline kolejka/judge/systems/local.py]: Unlikely (?) race condition with process. We should wait with process_start until memory reservation is acknowledged?

**Review by grzegorz-gutowski — COMMENTED** [review]
We should consider some synchronization between memory_reservation and process_start.

File: judges/cuda/test.yaml [hunk 2]
```diff
@@ -16,6 +16,7 @@ kolejka:
         storage     : '2G'
         workspace   : '2G'
         gpus        : 1
+        memory_gpu  : '128MB'
```
Comment (grzegorz-gutowski) [inline judges/cuda/test.yaml:19]: This should be larger. Judge default is 1G, example is 256MB.

File: kolejka/judge/systems/base.py [hunk 3]
```diff
@@ -336,7 +336,7 @@ def finalize():
         return result
 
     def execute_safe_command(self, command, stdin_path, stdout_path, stdout_append, stdout_max_bytes, stderr_path, stderr_append, stderr_max_bytes, environment, work_path, user, group, limits, result):
-        return self.execute_command(self, command, stdin_path, stdout_path, stdout_append, stdout_max_bytes, stderr_path, stderr_append, stderr_max_bytes, environment, work_path, user, group, limits, result)
+        return self.execute_command(command, stdin_path, stdout_path, stdout_append, stdout_max_bytes, stderr_path, stderr_append, stderr_max_bytes, environment, work_path, user, group, limits, result)
```
Comment (grzegorz-gutowski) [inline kolejka/judge/systems/base.py:339]: WOW. Scary!

File: kolejka/judge/tasks/base.py [hunk 4]
```diff
@@ -17,7 +17,7 @@ def __dir__():
 
 
 class TaskBase(AbstractTask):
-    def __init__(self, name=None, system=None, work_directory=None, environment=None, user=None, group=None, limits=None, limit_cpu_time=None, limit_real_time=None, limit_memory=None, limit_gpu_time=None, limit_gpu_memory=None, limit_cores=None, limit_pids=None, verbose=None, default_logs=None, result_on_error=None, result_on_time=None, result_on_memory=None, record_result=True, obligatory=False, safe=None):
+    def __init__(self, name=None, system=None, work_directory=None, environment=None, user=None, group=None, limits=None, limit_cpu_time=None, limit_real_time=None, limit_memory=None, limit_gpu_time=None, limit_gpu_memory=None, limit_cores=None, limit_pids=None, verbose=None, default_logs=None, result_on_error=None, result_on_time=None, result_on_memory=None, record_result=True, obligatory=False, safe=None, **kwargs):
```
Comment (grzegorz-gutowski) [inline kolejka/judge/tasks/base.py:20]: Why?

Comment (Raalsky) [reply kolejka/judge/tasks/base.py:20]: Ohh. Debugging artifact :wink:. To be removed.

File: judges/cuda/test.yaml (hunk 2 above)
Comment (Raalsky) [reply judges/cuda/test.yaml:19]: Sure. It's part of debugging artifacts. I wasn't sure about the source of values in Task limit. It was done just to made them odd.

File: kolejka/judge/systems/base.py (hunk 3 above)
Comment (Raalsky) [reply kolejka/judge/systems/base.py:339]: :smirk:

File: kolejka/judge/systems/local.py (hunk 1 above)
Comment (Raalsky) [reply kolejka/judge/systems/local.py]: Good point. At first, I was struggling with the termination of a thread so just copy-pasting logic from monitor process and I forgot to rethink the overall situation. I've transformed the memory reservation thread into process (😢) due to the ability to explicitly terminate right after the main command.

**Comment by Raalsky** [comment]
This need to be tested with generic judge

**Comment by Raalsky** [comment]
@grzegorz-gutowski tested with `generic` task and fixed bug with uninitialized GPU memory limit

File: judges/cuda/test.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        storage     : '2G'
+        workspace   : '2G'
+        gpus        : 1
+        memory_gpu  : '256MB'
+    satori:
+        result:
+            execute_time_real : '/io/executor/run/real_time'
+            execute_time_cpu  : '/io/executor/run/cpu_time'
+            execute_memory    : '/io/executor/run/memory'
+            execute_time_gpu  : '/io/executor/run/gpu_time'
```
Comment (grzegorz-gutowski) [inline judges/cuda/test.yaml:25]: How about an example based on profiler metrics?

File: judges/cuda/judge.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    tool_time = parse_time('60s')
+    prepare_time = parse_time('5s')
+    source_size_limit = parse_memory(args.test.get('source_size', '100K'))
+    binary_size_limit = parse_memory(args.test.get('binary_size', '10M'))
+    compile_time = parse_time(args.test.get('compile_time', '30s'))
+    cpp_standard = args.test.get('cpp_standard', 'c++17')
+    time_limit = parse_time(args.test.get('time', '10s'))
+    memory_limit = parse_memory(args.test.get('memory', '1G'))
+    time_gpu_limit = parse_time(args.test.get('time_gpu', '10s'))
+    memory_gpu_limit = parse_memory(args.test.get('memory_gpu', '1G'))
```
Comment (grzegorz-gutowski) [inline judges/cuda/judge.py:22]: Remove, or switch to metrics?

File: judges/cuda/judge.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        tool_cpp_standard=cpp_standard,
+        generator_source=args.test.get('generator', None),
+        verifier_source=args.test.get('verifier', None),
+        hint_path=args.test.get('hint', None),
+        hinter_source=args.test.get('hinter', None),
+        checker_source=args.test.get('checker', None),
+        limit_cores=1,
+        limit_time=time_limit,
+        limit_memory=memory_limit,
+        limit_gpu_time=time_gpu_limit,
```
Comment (grzegorz-gutowski) [inline judges/cuda/judge.py:49]: Remove, or switch to metrics

File: judges/cuda/test.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
+memory:       512MB
+memory_gpu:   256MB
```
Comment (grzegorz-gutowski) [inline judges/cuda/test.yaml:2]: 1GB

File: judges/cuda/test.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        time        : '20s'
+        memory      : '4G'
+        swap        : 0
+        cpus        : 4
+        network     : false
+        pids        : 256
+        storage     : '2G'
+        workspace   : '2G'
+        gpus        : 1
+        memory_gpu  : '256MB'
```
Comment (grzegorz-gutowski) [inline judges/cuda/test.yaml:19]: 1GB

File: judges/cuda/test.yaml (hunk 5 above)
Comment (grzegorz-gutowski) [inline judges/cuda/test.yaml:25]: remove?

File: kolejka/judge/systems/local.py [hunk 10]
```diff
@@ -12,10 +12,11 @@
 import time
 import threading
 import traceback
+import multiprocessing
```
Comment (grzegorz-gutowski) [inline kolejka/judge/systems/local.py:15]: no longer needed?

File: kolejka/judge/tasks/cuda.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+                self.set_result(None, metric_name, f'{metric_value:.4f}')
+
+            return self.result
+
+        return super().execute()
+
+class SolutionExecutableCudaTask(ExecutableCudaTask, SolutionExecutableTask):
+    pass
+
+class ProgramCudaTask(ProgramTask):
```
Comment (grzegorz-gutowski) [inline kolejka/judge/tasks/cuda.py:105]: Remove?

File: kolejka/judge/tasks/cuda.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    pass
+
+class SolutionProgramCudaTask(SolutionProgramTask):
```
Comment (grzegorz-gutowski) [inline kolejka/judge/tasks/cuda.py:108]: Remove?

File: kolejka/judge/tasks/cuda.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+    def solution_task_factory(self, **kwargs):
+        return SolutionExecutableCudaTask(
+            cuda_metrics=self._cuda_metrics,
+            cuda_profile_log=self._cuda_profile_log,
+            cuda_metrics_output=self._cuda_metrics_output,
+            **kwargs
+        )
+
+class MultipleIOCudaTask(MultipleIOTask):
```
Comment (grzegorz-gutowski) [inline kolejka/judge/tasks/cuda.py:137]: Remove?

File: judges/cuda/judge.py (hunk 6 above)
Comment (Raalsky) [reply judges/cuda/judge.py:22]: Is it about `time_gpu_limit` right? The line selected is with the `memory_gpu_limit` but I think it should be line above?

File: judges/cuda/test.yaml (hunk 5 above)
Comment (Raalsky) [reply judges/cuda/test.yaml]: I'll change it as staded in another comment -> try to get metric from profiler

**Pull request merged**